lib/CodeGen/CGDynamicCheck.cpp

```diff
@@ -111,6 +111,76 @@ void CodeGenFunction::EmitDynamicBoundsCheck(const Address PtrAddr, const Bounds
   EmitDynamicCheckBlocks(Builder.CreateAnd(LowerChk, UpperChk, "_Dynamic_check.range"));
 }
 
+void CodeGenFunction::EmitDynamicBoundsCheck(const Address BaseAddr,
+                                             const BoundsExpr *CastBounds,
+                                             const BoundsExpr *SubExprBounds) {
+  if (!getLangOpts().CheckedC)
+    return;
+
+  if (!SubExprBounds || !CastBounds)
+    return;
+
+  if (SubExprBounds->isAny() || SubExprBounds->isInvalid())
+    return;
+
+  // SubExprBounds can be Any by inference but CastBounds can't be Any
+  assert(!CastBounds->isAny());
+  if (CastBounds->isInvalid())
+    return;
+
+  // We can only generate the check if we have the bounds as a range.
+  if (!isa<RangeBoundsExpr>(SubExprBounds) ||
+      !isa<RangeBoundsExpr>(CastBounds)) {
+    llvm_unreachable(
+        "Can Only Emit Dynamic Bounds Check For RangeBounds Exprs");
+    return;
+  }
+
+  const RangeBoundsExpr *SubRange = dyn_cast<RangeBoundsExpr>(SubExprBounds);
+  const RangeBoundsExpr *CastRange = dyn_cast<RangeBoundsExpr>(CastBounds);
+
+  ++NumDynamicChecksRange;
+
+  // SubRange - bounds(lb, ub) vs CastRange - bounds(castlb, castub)
+  // Dynamic_check(lb <= castlb && castub <= ub)
+
+  // Emit the code to generate the pointer values
+  Address Lower = EmitPointerWithAlignment(SubRange->getLowerExpr());
```

Comment:
My understanding is that LLVM IR is a linearized sequence of instructions.   It is not an expression tree.   I believe that means that a sequence of instructions will be emitted here to compute the lower bound.   It is incorrect to compute the lower bound (or the upper bound) if BaseAddr is null.     The problem is the lower bound computation could fail at runtime if it involves a null pointer.   We must check that BaseAddr is not null before doing any of these computations.